### PR TITLE
[MLX] Support radix cache

### PR DESCRIPTION
**Pull request by yeahdongcn**
## Motivation

This PR adds a radix cache to the MLX backend, bringing it closer to feature parity with the PyTorch-based implementation and improving prefill throughput for workloads with shared prefixes.

## Modifications

Added a `kv_cache/` sub-package under the MLX backend with the core radix-cache components: a flat KV pool (`MlxKVPool`) for MLX-native data storage (auto-sized from available Metal memory), two cache flavors for prefill (`PoolBackedCache`, zero-copy gather) and decode (`ContiguousKVCache`, slice-assignment), and a batched attention wrapper (`MLXAttentionWrapper`) for BS>1 decode with vectorized RoPE.

Prefix-trie lookup, eviction, and slot allocation are handled by the scheduler's existing `RadixCache` and `TokenToKVPoolAllocator` — the MLX runner receives `prefix_slot_ids` and `new_slot_ids` from the scheduler and uses `MlxKVPool` purely as a data store.

Updated `model_runner.py` to wire scheduler-provided prefix/slot info, pool management, and deferred sync into the prefill/decode lifecycle. Updated `tp_worker.py` to extract `prefix_indices`, `out_cache_loc`, and `req_pool_idx` from `ModelWorkerBatch` and pass them to the MLX runner. Updated `bench_one_batch.py` to pass `disable_radix_cache`, `mem_fraction_static`, and `pool_size` for benchmarking.

### Logs

```bash
# First request
[2026-03-27 11:40:03] INFO:     127.0.0.1:56520 - "POST /v1/chat/completions HTTP/1.1" 200 OK
[2026-03-27 11:40:08] Prefill batch, #new-seq: 1, #new-token: 3662, #cached-token: 0, token usage: 0.09, #running-req: 0, #queue-req: 0, cuda graph: False, input throughput (token/s): 0.01

# Second request with the same prompt
[2026-03-27 11:40:42] INFO:     127.0.0.1:56520 - "POST /v1/chat/completions HTTP/1.1" 200 OK
[2026-03-27 11:40:44] Prefill batch, #new-seq: 1, #new-token: 1, #cached-token: 3661, token usage: 0.09, #running-req: 0, #queue-req: 0, cuda graph: False, input throughput (token/s): 102.45
```

## Benchmarking and Profiling

Model: Qwen/Qwen3-0.6B

### bs=1 input-len=60 output-len=100
```bash
Warmup ...
Prefill. latency: 0.12373 s, throughput:    484.91 token/s
Decode 0. Batch size: 1, latency: 0.02902 s, throughput:     34.46 token/s
Decode 1. Batch size: 1, latency: 0.02970 s, throughput:     33.67 token/s
Decode 2. Batch size: 1, latency: 0.02997 s, throughput:     33.37 token/s
Decode 3. Batch size: 1, latency: 0.02590 s, throughput:     38.61 token/s
Decode 4. Batch size: 1, latency: 0.02596 s, throughput:     38.53 token/s
Decode.  median latency: 0.02463 s, median throughput:     40.60 token/s
Total. latency:  0.907 s, throughput:    101.43 token/s
Benchmark ...
Prefill. latency: 0.05981 s, throughput:   1003.24 token/s
Decode 0. Batch size: 1, latency: 0.02457 s, throughput:     40.70 token/s
Decode 1. Batch size: 1, latency: 0.02656 s, throughput:     37.66 token/s
Decode 2. Batch size: 1, latency: 0.02859 s, throughput:     34.97 token/s
Decode 3. Batch size: 1, latency: 0.03229 s, throughput:     30.97 token/s
Decode 4. Batch size: 1, latency: 0.02955 s, throughput:     33.84 token/s
Decode.  median latency: 0.02510 s, median throughput:     39.85 token/s
Total. latency:  2.547 s, throughput:     62.83 token/s
```

### bs=4 input-len=60 output-len=100
```bash
Warmup ...
Prefill. latency: 0.35028 s, throughput:    685.16 token/s
Decode 0. Batch size: 4, latency: 0.05203 s, throughput:     76.88 token/s
Decode 1. Batch size: 4, latency: 0.04963 s, throughput:     80.60 token/s
Decode 2. Batch size: 4, latency: 0.05592 s, throughput:     71.54 token/s
Decode 3. Batch size: 4, latency: 0.05310 s, throughput:     75.34 token/s
Decode 4. Batch size: 4, latency: 0.05279 s, throughput:     75.77 token/s
Decode.  median latency: 0.05212 s, median throughput:     76.75 token/s
Total. latency:  1.951 s, throughput:    188.63 token/s
Benchmark ...
Prefill. latency: 0.23874 s, throughput:   1005.29 token/s
Decode 0. Batch size: 4, latency: 0.05667 s, throughput:     70.59 token/s
Decode 1. Batch size: 4, latency: 0.05219 s, throughput:     76.64 token/s
Decode 2. Batch size: 4, latency: 0.05203 s, throughput:     76.88 token/s
Decode 3. Batch size: 4, latency: 0.05087 s, throughput:     78.63 token/s
Decode 4. Batch size: 4, latency: 0.05011 s, throughput:     79.83 token/s
Decode.  median latency: 0.05360 s, median throughput:     74.63 token/s
Total. latency:  5.544 s, throughput:    115.44 token/s
```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/hardware_backend/mlx/kv_cache/attention_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        object.__setattr__(self, "_inner", inner)
+        object.__setattr__(self, "_layer_idx", layer_idx)
+
+    def __call__(self, x: mx.array, mask: Any = None, cache: Any = None) -> mx.array:
+        ctx = get_context()
+        if ctx is None:
+            return self._inner(x, mask=mask, cache=cache)
+        return self._batched_decode(x, ctx)
+
+    def _batched_decode(self, x: mx.array, ctx: BatchedDecodeContext) -> mx.array:
```
Comment (jonahbernard) [inline python/sglang/srt/hardware_backend/mlx/kv_cache/attention_wrapper.py:57]: Can you please add a unit test for this _batched_decode function?

**Review by jonahbernard — CHANGES_REQUESTED** [review]

File: python/sglang/srt/hardware_backend/mlx/kv_cache/radix_trie.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+        slot_ids: list[int],
+        prefix_len: int,
+        last_node: TrieNode,
+    ):
+        self.slot_ids = slot_ids
+        self.prefix_len = prefix_len
+        self.last_node = last_node
+
+
+class MlxRadixTrie:
```
Comment (jonahbernard) [inline python/sglang/srt/hardware_backend/mlx/kv_cache/radix_trie.py]: Can you please add comprehensive test coverage for all major functions in this class and PR in general to have test coverage like SGLang has in `test_radix_cache_unit.py` for the general SGLang radix cache implementation

File: python/sglang/srt/hardware_backend/mlx/kv_cache/contiguous_cache.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+from typing import TYPE_CHECKING
+
+import mlx.core as mx
+
+if TYPE_CHECKING:
+    from sglang.srt.hardware_backend.mlx.kv_cache.kv_pool import MlxKVPool
+
+
+class OffsetCache:
```
Comment (jonahbernard) [inline python/sglang/srt/hardware_backend/mlx/kv_cache/contiguous_cache.py:13]: Is this class necessary? I ran it without passing it in and it worked fine. 

Image

The cache isn't necessary in the mlx-lm source code forward pass:

Image

**Review by jonahbernard — CHANGES_REQUESTED** [review]

File: python/sglang/srt/hardware_backend/mlx/kv_cache/contiguous_cache.py (hunk 3 above)
Comment (yeahdongcn) [reply python/sglang/srt/hardware_backend/mlx/kv_cache/contiguous_cache.py:13]: mlx-lm models expect `cache=` to be a list of per-layer cache objects. Before attention layers run, the model reads `cache[i].offset` for RoPE positioning and calls `cache[i].make_mask()` for attention masking. Passing `cache=None` would make RoPE assume position 0 and produce incorrect masks. `OffsetCache` is a data-free shim that satisfies this protocol—providing the correct offset and mask (without allocating any KV storage), since actual KV management is handled separately by `MLXAttentionWrapper` via `BatchedDecodeContext`.

File: python/sglang/srt/hardware_backend/mlx/kv_cache/attention_wrapper.py (hunk 1 above)
Comment (yeahdongcn) [reply python/sglang/srt/hardware_backend/mlx/kv_cache/attention_wrapper.py:57]: Yes, of course. I think we can add this in a separate PR once the Mac CI infrastructure is ready. I already have a few unit test files locally.

File: python/sglang/srt/hardware_backend/mlx/kv_cache/radix_trie.py (hunk 2 above)
Comment (yeahdongcn) [reply python/sglang/srt/hardware_backend/mlx/kv_cache/radix_trie.py]: Same as above.

**Comment by yeahdongcn** [comment]
@hnyls2002 @alexnails Ping 😄

File: python/sglang/srt/hardware_backend/mlx/kv_cache/attention_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+        attn_mask = None
+        seq_lens_plus1 = [s + 1 for s in ctx.seq_lens]
+        if min(seq_lens_plus1) < max_len:
+            positions = mx.arange(max_len)
+            valid_lens = mx.array(seq_lens_plus1, dtype=mx.int32)
+            mask_bool = positions[None, :] >= valid_lens[:, None]
+            attn_mask = mx.where(
+                mask_bool[:, None, None, :],
+                mx.array(-1e9, dtype=queries.dtype),
```
Comment (alexnails) [inline python/sglang/srt/hardware_backend/mlx/kv_cache/attention_wrapper.py]: can we finfo / numeric limits this?

File: python/sglang/srt/hardware_backend/mlx/model_runner.py [hunk 5]
```diff
@@ -1,103 +1,109 @@
...
+
     @staticmethod
     def _extract_logits(model_output):
         """Extract logits from model output, handling both tuple and direct returns."""
         if isinstance(model_output, tuple):
             return model_output[0]
         return model_output
 
+    def _acquire_cache(self) -> list[ContiguousKVCache]:
+        """Get a reusable cache list from the pool, or create a new one."""
```
Comment (alexnails) [inline python/sglang/srt/hardware_backend/mlx/model_runner.py:85]: do we flush after previous use? or just set offset?

File: python/sglang/srt/hardware_backend/mlx/kv_cache/attention_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+import threading
+from dataclasses import dataclass
+from typing import Any, Optional
+
+import mlx.core as mx
+import mlx.nn as nn
+
+from sglang.srt.hardware_backend.mlx.kv_cache.contiguous_cache import ContiguousKVCache
+
+_thread_local = threading.local()
```
Comment (alexnails) [inline python/sglang/srt/hardware_backend/mlx/kv_cache/attention_wrapper.py:14]: does this impl break the chained decode stuff @changminbark is working on ?

File: python/sglang/srt/hardware_backend/mlx/model_runner.py [hunk 7]
```diff
@@ -107,200 +113,336 @@ def _load_model(self):
...
+        if not layer_list:
+            raise RuntimeError("Cannot determine attention config: no layers found")
+        sample_attn = getattr(layer_list[0], attn_attr)
+        if isinstance(sample_attn, MLXAttentionWrapper):
+            sample_attn = sample_attn._inner
+        n_kv_heads = sample_attn.n_kv_heads
+        if hasattr(sample_attn, "head_dim"):
+            head_dim = sample_attn.head_dim
+        elif hasattr(sample_attn, "k_proj") and hasattr(sample_attn.k_proj, "weight"):
+            head_dim = sample_attn.k_proj.weight.shape[0] // n_kv_heads
```
Comment (alexnails) [inline python/sglang/srt/hardware_backend/mlx/model_runner.py:135]: is this always right? I feel like this does not work for DeepSeek style heads (e.g separating k and v)

File: python/sglang/srt/hardware_backend/mlx/kv_cache/attention_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+        # Vectorized RoPE with per-batch offsets
+        offsets = mx.array(ctx.seq_lens, dtype=mx.int32)
+        queries = inner.rope(queries, offset=offsets)
+        keys = inner.rope(keys, offset=offsets)
+
+        layer_caches = ctx.layer_caches[layer_idx]
+        max_len = max(ctx.seq_lens) + 1
+
+        # TODO: replace per-request loop with native batched/ragged
```
Comment (alexnails) [inline python/sglang/srt/hardware_backend/mlx/kv_cache/attention_wrapper.py:88]: can we still do batched prefill?

File: python/sglang/srt/hardware_backend/mlx/model_runner.py [hunk 9]
```diff
@@ -107,200 +113,336 @@ def _load_model(self):
...
+            )
+            set_context(ctx)
+            try:
+                max_offset = max(seq_lens)
+                shim_cache = [OffsetCache(offset=max_offset) for _ in range(num_layers)]
+                last_tokens = [self._req_token_ids[rid][-1] for rid in req_ids]
+                batched_input = mx.array(last_tokens, dtype=mx.int32)[:, None]
+                model_output = self.model(batched_input, cache=shim_cache)
+                logits = self._extract_logits(model_output)
+                next_tokens_mlx = mx.argmax(logits[:, -1, :], axis=-1)
```
Comment (alexnails) [inline python/sglang/srt/hardware_backend/mlx/model_runner.py:405]: greedy sampling only? or follow up later?

Comment (yeahdongcn) [reply python/sglang/srt/hardware_backend/mlx/model_runner.py:405]: Yes, the current MLX path is greedy-only for now. `prefill()`, `extend()`, and `decode_batch()` all take `argmax` directly from the logits, so this PR keeps scope focused on radix cache + batched decode plumbing rather than sampler parity. Non-greedy sampling should be a follow-up once the MLX backend's cache/data path is stable.

File: python/sglang/srt/hardware_backend/mlx/model_runner.py (hunk 7 above)
Comment (yeahdongcn) [reply python/sglang/srt/hardware_backend/mlx/model_runner.py:135]: Good catch. The current fallback is only intended for the attention layouts we support today in the MLX backend, where KV head dim can be inferred from `k_proj.weight.shape[0] // n_kv_heads`. It is not meant as a universally correct rule for every architecture. In particular, models with asymmetric or nonstandard K/V projection layouts may need model-specific handling here. I did not widen this PR to solve that generically, but I agree this should either be guarded more explicitly or handled in a follow-up when we add broader architecture coverage.

File: python/sglang/srt/hardware_backend/mlx/model_runner.py (hunk 5 above)
Comment (yeahdongcn) [reply python/sglang/srt/hardware_backend/mlx/model_runner.py:85]: We only reset `offset` when reusing a `ContiguousKVCache`. We do not explicitly zero the old buffer contents. That is safe here because all reads go through `get_kv()`, which slices only the live prefix `[:, :, :self.offset, :]`, and subsequent writes overwrite from offset 0 forward. So old data may remain in the backing buffer physically, but it is logically dead once `offset` is reset.

File: python/sglang/srt/hardware_backend/mlx/kv_cache/attention_wrapper.py (hunk 8 above)
Comment (yeahdongcn) [reply python/sglang/srt/hardware_backend/mlx/kv_cache/attention_wrapper.py:88]: Yes. This wrapper is only for batched decode. Batched prefill still goes through the normal MLX model forward path. The radix-cache prefill optimization lives in `MlxModelRunner.prefill()` / `extend()` and uses either `PoolBackedCache` or `ContiguousKVCache` there. So this change does not remove batched prefill support, it only adds a separate BS>1 decode path.

File: python/sglang/srt/hardware_backend/mlx/kv_cache/attention_wrapper.py (hunk 6 above)
Comment (yeahdongcn) [reply python/sglang/srt/hardware_backend/mlx/kv_cache/attention_wrapper.py:14]: I'm not entirely sure. There are a few new changes built on top of my branch, so we may need to do some rebase/merge work.

File: python/sglang/srt/hardware_backend/mlx/kv_cache/attention_wrapper.py (hunk 4 above)
Comment (yeahdongcn) [reply python/sglang/srt/hardware_backend/mlx/kv_cache/attention_wrapper.py]: I think your suggestion is aiming to remove the use of a `magic constant` here. However, in practice, we only need a sufficiently negative value such that `exp(masked_logit)` is effectively zero.

For example, if softmax sees logits like:

```
[2.1, 1.3, -1e9]
```

the masked entry already contributes ~0 probability. Using `-3.4e38` instead of `-1e9` doesn’t provide any meaningful benefit.

```python
>>> mx.finfo(mx.float32).min
-3.4028234663852886e+38
```

In fact, I'm not even sure `-1e9` is strictly necessary here 😂. I'll test locally to confirm.

Comment (yeahdongcn) [reply python/sglang/srt/hardware_backend/mlx/kv_cache/attention_wrapper.py]: Verified that `-1e4` still works. However, based on the discussion on Slack, I think there is a valid concern here that we can't reliably assume the range of these values, and this could lead to unexpected issues.

Comment (yeahdongcn) [reply python/sglang/srt/hardware_backend/mlx/kv_cache/attention_wrapper.py]: Fixed now.

**Review by alexnails — APPROVED** [review]
as this is self contained to MLX, LGTM. Just make sure CI passes and there are no build errors, etc

File: python/sglang/srt/hardware_backend/mlx/kv_cache/attention_wrapper.py (hunk 6 above)
Comment (changminbark) [reply python/sglang/srt/hardware_backend/mlx/kv_cache/attention_wrapper.py:14]: There's a good chance it may break it, so I will need to do some merge work and apply fixes.
